src/rules/package-lint-dependencies-conform.ts

```diff
@@ -0,0 +1,88 @@
+import { PackageManifestSchema, RuleName } from './types';
+import type { RuleExecutionFailure } from '../execute-rules';
+
+export default buildRule({
+  name: RuleName.PackageLintDependenciesConform,
+  description:
+    'Do the lint-related `devDependencies` in `package.json` conform?',
+  dependencies: [RuleName.RequireValidPackageManifest],
+  execute: async ({ project, template, pass, fail }) => {
+    const entryPath = 'package.json';
+
+    const templateManifest = await template.fs.readJsonFileAs(
+      entryPath,
+      PackageManifestSchema,
+    );
+
+    const projectManifest = await project.fs.readJsonFileAs(
+      entryPath,
+      PackageManifestSchema,
+    );
+
+    const failures: RuleExecutionFailure[] = await lintPackageConform(
+      templateManifest.devDependencies,
+      projectManifest.devDependencies,
+    );
+
+    return failures.length === 0 ? pass() : fail(failures);
+  },
+});
+
+/**
+ * Validates whether target project has all the required lint packages with versions matching with template project.
+ *
+ * @param templateDependencies - The record of lint package name and version from template.
+ * @param projectDependencies - The record of lint package name and version from project.
+ */
+async function lintPackageConform(
+  templateDependencies: Record<string, string>,
+  projectDependencies: Record<string, string>,
+): Promise<RuleExecutionFailure[]> {
+  const templateLintPackageNames = await getTemplateLintPackageNames(
+    templateDependencies,
+  );
+  const failures: RuleExecutionFailure[] = [];
+  for (const lintPackage of templateLintPackageNames) {
+    const projectPackageVersion = projectDependencies[lintPackage];
+    const templatePackageVersion = templateDependencies[lintPackage] as string;
+    if (!projectPackageVersion) {
+      failures.push({
+        message: `\`package.json\` should list \`"${lintPackage}": "${templatePackageVersion}"\` in \`devDependencies\`, but does not.`,
+      });
+
+      continue;
+    }
+
+    if (projectPackageVersion !== templatePackageVersion) {
+      failures.push({
+        message: `${lintPackage} version is "${projectPackageVersion}", when it should be "${templatePackageVersion}".`,
+      });
+    }
+  }
+
+  return failures;
+}
+
+/**
+ * Extracts list of lint package names from template's package.json.
+ *
+ * @param templateDependencies - The record of lint package name and version.
+ */
+async function getTemplateLintPackageNames(
+  templateDependencies: Record<string, string>,
+): Promise<string[]> {
+  const requiredPackagePatterns: RegExp[] = [
+    /^@metamask\/eslint-config-[^/]+$/u,
+    /^@typescript-eslint\/[^/]+$/u,
+    /^eslint(-[^/]+)?$/u,
+    /^prettier(-[^/]+)?$/u,
+  ];
+
+  const templatePackageNames = Object.keys(templateDependencies);
+  return templatePackageNames.filter((packageName) => {
+    return requiredPackagePatterns.some((pattern) => {
+      return packageName.match(pattern);
+    });
+  });
+}
```

Comment:
To extend a previous comment, we could have this function return a slice of the object rather than just the properties, and that would allow us to iterate over its properties and values in tandem:

``` suggestion
function getTemplateLintPackages(
  templateDependencies: Record<string, string>,
): Record<string, string> {
  const requiredPackagePatterns: RegExp[] = [
    /^@metamask\/eslint-config-[^/]+$/u,
    /^@typescript-eslint\/[^/]+$/u,
    /^eslint(-[^/]+)?$/u,
    /^prettier(-[^/]+)?$/u,
  ];

  return Object.entries(templateDependencies).reduce<Record<string, string>>((packages, [packageName, packageVersion]) => {
    if (requiredPackagePatterns.some((pattern) => packageName.match(pattern)) {
      return { ...packages, [packageName]: packageVersion };
    }
    return packages;
  }, {});
}